elasticsearch/src/main/java/org/hibernate/search/elasticsearch/client/impl/BulkRequest.java

```diff
@@ -78,6 +83,45 @@ public void execute() {
 		}
 	}
 
+	/**
+	 * Call every result handlers, managing exceptions so that one failing handler will not prevent
+	 * others from being called.
+	 *
+	 * <p>No locally caught exception will be re-thrown, but the "main" exception will be returned.
+	 *
+	 * <p>The "main" exception is either the {@code preexistingMainException} parameter (if non-null) or the first
+	 * locally caught exception. Every locally caught exception that is not the "main" exception will be
+	 * added to the "main" exception.
+	 *
+	 * @param successfulItems The requests and their results, for calling result handler.
+	 * @param preexistingMainException The exception to use as a "main" exception (see above), or {@code null}.
+	 * @return The "main" exception (see above), or {@code null} if there isn't any.
+	 */
+	private RuntimeException callHandlers(Map<BackendRequest<?>, BulkResultItem> successfulItems, RuntimeException preexistingMainException) {
+		RuntimeException mainException = preexistingMainException;
+		for ( Map.Entry<BackendRequest<?>, BulkResultItem> entry : successfulItems.entrySet() ) {
```

Comment:
The `IndexingMonitor` can become a contention point when it's invoked too frequently, for example when running on servers with many CPUs, one `MassIndexer` job will use them all in parallel, and the pattern is relatively effective because it attempts to not share much; the monitor is one of the exceptions.

For this reason it takes a `long` argument, so that its counters can be incremented by values larger than one; so I would expect a Bulk Request not cause invoking the `IndexingMonitor#documentsAdded( 1L )` method N times, but invoke it with a paremeter N.

On the other hand, I like your idea of having the `handle` method to accept the `BulkResultItem` argument. 
Could you move this loop into the `BackendRequestSuccessHandler` implementation? There it can inspect the bulk result, and invoke the `IndexingMonitor` once as necessary.